### PR TITLE
Generate conversions/deep-copies for experimental.

### PR DESCRIPTION
**Pull request by uluyol**
This addresses the first issue discussed #12160 enough to unblock using the experimental api. 

This PR will generate and register duplicate functions but this is benign, deduplicating is being tackled in PR #12228.

@smarterclayton @nikhiljindal @wojtek-t

### CODE REVIEW

**Comment by caesarxuchao** [comment]
sub/

**Comment by wojtek-t** [comment]
generally LGTM

File: hack/update-generated-deep-copies.sh [hunk 1]
```diff
@@ -25,51 +25,51 @@ kube::golang::setup_env
...
-  apiVersions=""
-  for ver in ${versions}; do
-    # Ensure that the version being processed is registered by setting
-    # KUBE_API_VERSIONS.
-    if [ "${ver}" != "api" ]; then
-      apiVersions="${ver}"
-    fi
-    KUBE_API_VERSIONS="${apiVersions}" generate_version "${ver}"
-  done
+	local versions="api/ api/v1 expapi/ expapi/v1"
```
Comment (smarterclayton) [inline hack/update-generated-deep-copies.sh]: Can we make this a variable that can be specified by the caller but defaults to these values?  If so, this script is then reusable by other projects (openshift) and callers.

**Comment by caesarxuchao** [comment]
Thank you. LGTM. I will apply the label after you resolve the comment with Clayton.

File: hack/update-generated-deep-copies.sh (hunk 1 above)
Comment (uluyol) [inline hack/update-generated-deep-copies.sh]: Done. Just do VERSIONS="g/v g/v2" hack/update-generated-deep-copies.sh (or update-generated-conversions.sh).

**Comment by uluyol** [comment]
Pushed a new version that defaults the versions.

**Comment by smarterclayton** [comment]
LGTM

**Comment by caesarxuchao** [comment]
Shippable fails:
--- FAIL: TestNoManualChangesToGenerateConversions (0.34s)
conversion_generation_test.go:132: please update conversion functions; generated: v1.functions.txt; first diff:
expected: api "k8s.io/kubernetes/pkg/api"
got: reflect "reflect"
--- FAIL: TestNoManualChangesToGenerateDeepCopies (0.34s)
conversion_generation_test.go:132: please update conversion functions; generated: api.deep_copy.txt; first diff:
expected: resource "k8s.io/kubernetes/pkg/api/resource"
got: time "time"
conversion_generation_test.go:132: please update conversion functions; generated: v1.deep_copy.txt; first diff:
expected: api "k8s.io/kubernetes/pkg/api"
got: time "time"
FAIL
coverage: 79.1% of statements
FAIL	k8s.io/kubernetes/pkg/runtime	1.244s

**Comment by uluyol** [comment]
Pushed a commit to remove the problematic tests. We already run hack/verify-generated-{conversions,deep-copies}.sh which check the same things.

**Comment by uluyol** [comment]
I'm guessing that shippable is failing because a temporary file was pushed (see #12504).

**Comment by uluyol** [comment]
Rebased, hopefully tests pass now.

**Comment by uluyol** [comment]
I have no idea what's going on with shippable, but I don't think it's this PR. Please review the the latest changes.

**Comment by caesarxuchao** [comment]
The changes LGTM. But the Shippable error is stable, and other Shippable builds don't suffer from the same problem, so I think we need to fix Shippable before merge.

**Comment by uluyol** [comment]
I forgot to update verify-generated-*.sh. The conversions/deep copies were old which was likely causing the dirty git state (and the verification scripts didn't check the experimental api).

Verification scripts should now be checking expapi and updated conversions/deep copies have been pushed.

**Comment by caesarxuchao** [comment]
LGTM. Thank you!

**Comment by cjcullen** [comment]
@uluyol Can you rebase and re-push?

**Comment by uluyol** [comment]
Rebased.

**Comment by eparis** [comment]
/me grumbles that now I need to build and install `goimports` before I can do a git commit...

**Comment by derekwaynecarr** [comment]
We should enumerate all the prerequisite tools needed to do a commit in our
developer guide.  I am not sure we actually have a definitive guide

**Comment by eparis** [comment]
actually, can we just use golang.org/x/tools/imports Process() instead of making a whole new command line flag?

File: hack/after-build/update-generated-deep-copies.sh [hunk 2]
```diff
@@ -53,27 +56,24 @@ EOF
...
-  apiVersions=""
-  for ver in ${versions}; do
-    # Ensure that the version being processed is registered by setting
-    # KUBE_API_VERSIONS.
-    if [ "${ver}" != "api" ]; then
-      apiVersions="${ver}"
-    fi
-    KUBE_API_VERSIONS="${apiVersions}" generate_version "${ver}"
-  done
+	local versions="$@"
```
Comment (eparis) [inline hack/after-build/update-generated-deep-copies.sh]: This broken the {verify/update}-generated-deep-copies.sh function as versions is empty pretty much always.  I'm assuming you thought about this with DEFAULT_VERSIONS and VERSIONS below, but you didn't plumb those through...

File: hack/after-build/update-generated-conversions.sh [hunk 3]
```diff
@@ -44,14 +44,19 @@ EOF
 // AUTO-GENERATED FUNCTIONS END HERE
 EOF
 
-	mv "$TMPFILE" "pkg/api/${version}/conversion_generated.go"
+	env GOPATH=$(godep path):$GOPATH goimports -w "$TMPFILE"
```
Comment (eparis) [inline hack/after-build/update-generated-conversions.sh]: Why are you setting GOPATH at all?  Shouldn't this be taken care of by `kube::golang::setup_env()`  I certainly don't want random stuff I have outside of the tree somehow messing with these commands.

File: hack/after-build/update-generated-deep-copies.sh (hunk 2 above)
Comment (uluyol) [inline hack/after-build/update-generated-deep-copies.sh]: Agh, lost that in the rebase. Will fix.

File: hack/after-build/update-generated-conversions.sh (hunk 3 above)
Comment (uluyol) [inline hack/after-build/update-generated-conversions.sh]: Was not aware of that. Will fix in a followup PR.

**Pull request merged**